Enhancements/kinetics units

bpatel2107 opened the pull request:
I've added in units to the `Kinetics` object and then modified the `Species` and `LocalSpecies` to take units from `Kinetics` directly. The object needs to be populated with data in the following format as indicated by the docstriing

```python
"""
    - psi_n       [units] dimensionless
    - r/a         [units] dimensionless
    - Charge      [units] elementary_charge
    - Mass        [units] kg
    - Temperature [units] eV
    - Density     [units] meter**-3
    - Rotation    [units] meter/second
"""
 ```

bpatel2107 committed:
Added in rhoref as a normalisation unit

bpatel2107 committed:
Modified GKOutputReader.py to use DatasetWrapper and take in local_norm as argument

bpatel2107 committed:
Added units to GS2 output coords

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Changed base GKOutputReader.py to read field as if separate

bpatel2107 committed:
Restructured fields in GS2 output to be separate values

bpatel2107 committed:
Merge remote-tracking branch 'origin/enhancements/gk_output_normalisation' into enhancements/gk_output_normalisation

# Conflicts:
#	pyrokinetics/gk_code/GKOutputReader.py
#	pyrokinetics/gk_code/GKOutputReaderGS2.py

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Added qref and made a dict of references needed for pint-xarray conversions

bpatel2107 committed:
Black format and add in pint dequantifies

bpatel2107 committed:
Add in field/flux units and convert from GS2 to pyro convention

bpatel2107 committed:
Merge remote-tracking branch 'origin/enhancements/gk_output_normalisation' into enhancements/gk_output_normalisation

# Conflicts:
#	pyrokinetics/gk_code/GKOutputReaderGS2.py

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Made field/flux units a method from base GKOutputReader.py class

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Added units to kinetics objects and modfied object that use Kinetics

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Update docstring for Kinetics

ZedThree commented on pyrokinetics/constants.py:
```diff
@@ -1,15 +1,16 @@
 eps0 = constants.epsilon_0
 
-electron_charge = constants.elementary_charge
+electron_charge = constants.elementary_charge * units.elementary_charge
```
I think `units.elementary_charge` is the value + units -- worth double checking at least. I think electron and proton mass are in there too. In fact, we could probably just use those directly?

bpatel2107 replied:
It looks like it is just the value. Not sure scipy supports pint units yet

bpatel2107 committed:
Add units into length scales a_lt -> inverse_lt

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Merge branch 'unstable' into enhancements/kinetics_units

bpatel2107 commented:
I've added in units to the kinetic gradients such that it would now be possible to change from $a/L_T$ to $R_{maj}/L_T$ using pint. I thought keeping the variable name as `local_species.electron.a_lt` didn't quite right as this term now has units of `1/lref` so I changed it to `inverse_lt` which will break existing code. We seem to be going through some major changes so now seems like a good time to do this.

@arkabokshi FYI

ZedThree commented:
If we're changing the names, is it worth changing straight to the IMAS names? They call them `density_log_gradient_norm` etc. Bit of a mouthful, but that's what tab-completion is for :)

bpatel2107 commented:
Well the question then becomes do we just change everything?

bpatel2107 committed:
Modify UnitSpline back to unstable version

bpatel2107 committed:
[skip ci] Apply black changes

bpatel2107 committed:
Trigger build

bpatel2107 commented:
Happy to discuss this further as it may be useful in the long run! Will probably keep this as a separate PR though (mentioned in #10 as a reminder) so but this is probably a separate issue for now. Given that the code is going to change quite a bit over the next few months I think its best to get these changes in so we don't have lots of open PRs lying around
